ksqldb-engine/src/main/java/io/confluent/ksql/format/DefaultFormatInjector.java

```diff
@@ -0,0 +1,190 @@
+    }
+  }
+
+  private Optional<ConfiguredStatement<CreateSource>> injectForCreateStatement(
+      final ConfiguredStatement<CreateSource> original
+  ) {
+
+    final CreateSource statement = original.getStatement();
+    final CreateSourceProperties properties = statement.getProperties();
+
+    final Optional<FormatInfo> keyFormat = properties.getKeyFormat();
+    final Optional<FormatInfo> valueFormat = properties.getValueFormat();
+
+    if (keyFormat.isPresent() && valueFormat.isPresent()) {
+      return Optional.empty();
+    }
+
+    final KsqlConfig config = getConfig(original);
+    validateConfig(config, keyFormat, valueFormat);
+
+    final CreateSourceProperties injectedProps = properties.withFormats(
+        keyFormat.map(FormatInfo::getFormat)
+            .orElse(config.getString(KsqlConfig.KSQL_DEFAULT_KEY_FORMAT_CONFIG)),
+        valueFormat.map(FormatInfo::getFormat)
+            .orElse(config.getString(KsqlConfig.KSQL_DEFAULT_VALUE_FORMAT_CONFIG))
+    );
+    final CreateSource withFormats = statement.copyWith(
+        original.getStatement().getElements(),
+        injectedProps
+    );
+
+    final PreparedStatement<CreateSource> prepared = buildPreparedStatement(withFormats);
+    final ConfiguredStatement<CreateSource> configured = ConfiguredStatement
+        .of(prepared, original.getConfigOverrides(), original.getConfig());
+
+    return Optional.of(configured);
+  }
+
+  private static void validateConfig(
+      final KsqlConfig config,
+      final Optional<FormatInfo> keyFormat,
+      final Optional<FormatInfo> valueFormat
+  ) {
+    if (!keyFormat.isPresent()
+        && config.getString(KsqlConfig.KSQL_DEFAULT_KEY_FORMAT_CONFIG) == null) {
+      throw new KsqlException("Statement is missing the '"
+          + CommonCreateConfigs.KEY_FORMAT_PROPERTY + "' property from the WITH clause. "
+          + "Either provide one or set a default via the '"
+          + KsqlConfig.KSQL_DEFAULT_KEY_FORMAT_CONFIG + "' config.");
+    }
+    if (!valueFormat.isPresent()
+        && config.getString(KsqlConfig.KSQL_DEFAULT_VALUE_FORMAT_CONFIG) == null) {
+      throw new KsqlException("Statement is missing the '"
+          + CommonCreateConfigs.VALUE_FORMAT_PROPERTY + "' property from the WITH clause. "
+          + "Either provide one or set a default via the '"
+          + KsqlConfig.KSQL_DEFAULT_VALUE_FORMAT_CONFIG + "' config.");
+    }
+  }
+
+  private static KsqlConfig getConfig(final ConfiguredStatement<?> statement) {
+    return statement.getConfig().cloneWithPropertyOverwrite(statement.getConfigOverrides());
+  }
+
+  private static void validateLegacyFormatProperties(final ConfiguredStatement<?> statement) {
+    if (statement.getStatement() instanceof CreateSource) {
+      final CreateSource createStatement = (CreateSource) statement.getStatement();
+
+      if (createStatement.getProperties().getKeyFormat().isPresent()) {
+        throwKeyFormatDisabled(statement.getStatementText());
+      }
```

Comment:
nit: 

Rather than having a method `throwKeyFormatDisabled ` that always throws an exception, I'd suggest having a `keyFormatDisabledException` method that creates the exception:

```suggestion
      if (createStatement.getProperties().getKeyFormat().isPresent()) {
        throw keyFormatDisabledException(statement.getStatementText());
      }
```

The benefit, is that it's easier to grok as you know by looking at the code that an exception is always thrown, without having to go check another function. This pattern can also avoid compiler issues, e.g. warnings and errors about code paths not initializing variables.

Reply:
Cool tip! I'm sold. 

Unfortunately these methods were removed as part of the change to allow the use of `KEY_FORMAT` and `FORMAT` even if the feature flag is disabled so we won't be able to realize the shinier v2 of this code ;o